pkg/daemon/writer.go

```diff
@@ -126,24 +127,42 @@ func (nw *clusterNodeWriter) Run(stop <-chan struct{}) {
 		case <-stop:
 			return
 		case msg := <-nw.writer:
-			r := implSetNodeAnnotations(nw.client, nw.lister, nw.nodeName, msg.annos)
+			r := implSetNodeAnnotations(nw.client, nw.lister, nw.nodeName, msg.annos, msg.annosToDelete)
 			msg.responseChannel <- r
 		}
 	}
 }
 
 // SetDone sets the state to Done.
-func (nw *clusterNodeWriter) SetDone(dcAnnotation string) error {
+func (nw *clusterNodeWriter) SetDone(state *stateAndConfigs) error {
 	annos := map[string]string{
 		constants.MachineConfigDaemonStateAnnotationKey: constants.MachineConfigDaemonStateDone,
-		constants.CurrentMachineConfigAnnotationKey:     dcAnnotation,
+		constants.CurrentMachineConfigAnnotationKey:     state.currentConfig.GetName(),
```

Comment:
Just to double check, this is going from taking the desired annotation to instead our currentconfig state, which at this point should have been updated, right? Any potential that this would not be the case here?

Reply:
That is a great point, thank you for calling that out. It should have been obvious from the variable name that is what was happening, although I like that this particular change makes that even more explicit. I will adjust this to take the `desiredConfig` and `desiredImage` instead. Nice catch!

Reply:
Did a bit of testing with setting this to `desiredConfig` and `desiredImage` and overall, it seemed to be fine. But I wanted to explain why this worked in its current state to address any confusion around this:

- After we've applied the update, we write the "new config" and "new image" to the nodes' filesystem in `/etc/machine-config-daemon/currentconfig` and `/etc/machine-config-daemon/currentimage`, respectively. In this case, the "new config" and "new image" are the `desiredConfig` and `desiredImage`. We then reboot the node (for updates that require it, anyway).
- After we reboot, when we call `getCurrentConfigOnDisk()`, the values it loads become `currentConfig` and `currentImage`, respectively.
- `SetDone()` gets called after the node reboot (for updates that require it). Because of the above point, `SetDone()` gets called with the current on disk config and image.

With this in mind, I'll leave this as-is for now and rebase. That said, it could be worth our while to address the confusion in the future.